### PR TITLE
fix: live captions don't show up in consistent manner

### PR DESCRIPTION
**Pull request by VisheshMittal**
## Description
This is a fix for live captions, i.e. during LIVE streaming, subtitles used to show up in an inconsistent fashion. 
If there's any "DISCONTINUITY" tag in the subtitle playlist, and if the discontinuities for subtitle playlist are not synchronized/aligned with the audio, video playlists then we observe that the subtitles either don't show up at all, or stop showing after some time. This PR addreses the same issue
Linked issues: https://github.com/videojs/video.js/issues/6353

## Specific Changes proposed
For LIVE/VOD workflows, the discontinuity information (segment with discontinuity) might not be consistent across media-playlists(audio/video/vtt). So, for segment loading decisions, we're leveraging "lastBufferedSegmentTimestamp". Based on the start-time of the segment last buffered by audio/video segment loaders, vtt-segment-loader decides whether to loading future segments or not, after comparing start-time values.

### CODE REVIEW

File: src/sync-controller.js [hunk 1]
```diff
@@ -382,11 +383,12 @@ export default class SyncController extends videojs.EventTarget {
     }
   }
 
-  timestampOffsetForTimeline(timeline) {
-    if (typeof this.timelines[timeline] === 'undefined') {
+  timestampOffsetForSegment(segmentInfo) {
+    if (segmentInfo.timeline > 0 &&
```
Comment (DaveStein) [inline src/sync-controller.js]: You might want to use intermediary variables for clarity. I'm not sure what these checks are doing.

```
const somethingDescriptive = (!segmentInfo.startOfSegment || segmentInfo.startOfSegment > this.lastBufferedSegmentTimestamp);

if (segmentInfo.timeline > 0 && somethingDescriptive) {
  return null;
}

Comment (VisheshMittal) [reply src/sync-controller.js]: Done

**Comment by gesinger** [comment]
Hey @VisheshMittal , thank you for the PR! Especially the documentation, we really appreciate it.

We were wondering though if you had a sample of your content. Based on the description, the content does not appear to be spec compliant.

A subtitle playlist is considered an alternative rendition, which must meet the constraints for a variant stream, which includes the requirement that: "Matching content in Variant Streams MUST have matching Discontinuity Sequence Numbers"

See:
https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-4.3.4.2.1
https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-6.2.4

But if you have some sample content to take a look at, it may help us in determining an appropriate resolution.

Thanks again!

**Comment by VisheshMittal** [comment]
> Hey @VisheshMittal , thank you for the PR! Especially the documentation, we really appreciate it.
> 
> We were wondering though if you had a sample of your content. Based on the description, the content does not appear to be spec compliant.
> 
> A subtitle playlist is considered an alternative rendition, which must meet the constraints for a variant stream, which includes the requirement that: "Matching content in Variant Streams MUST have matching Discontinuity Sequence Numbers"
> 
> See:
> https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-4.3.4.2.1
> https://tools.ietf.org/html/draft-pantos-http-live-streaming-23#section-6.2.4
> 
> But if you have some sample content to take a look at, it may help us in determining an appropriate resolution.
> 
> Thanks again!

Hi @gesinger ,
Thanks for your response, and the spec links. As of now, the live link (where the behaviour is visible) that i'm using, gets expired as soon as the livestream ends. I've already mentioned the contents of the sample manifest within the [`sync-controller`](https://github.com/videojs/http-streaming/blob/ce9920395c6bc16742a82525b5ac7d1bbc17fac2/docs/sync-controller.md) doc. I've also attached the contents of the sample below for viewing.

Please let me know, if you'd need a long-lived manifest url so as to check the behaviour. It'd take some work on my end, but i should be able to come up with it after some changes.

#### Manifest(video,format=m3u8-aapl)
```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-ALLOW-CACHE:NO
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-TARGETDURATION:17
#EXT-X-PROGRAM-DATE-TIME:2020-02-12T09:52:26Z
#EXTINF:16.666000,no-desc
Fragments(video=0,format=m3u8-aapl)
#EXTINF:16.666000,no-desc
Fragments(video=1499940,format=m3u8-aapl)
#EXTINF:16.667000,no-desc
Fragments(video=2999880,format=m3u8-aapl)
#EXTINF:16.667000,no-desc
Fragments(video=4499910,format=m3u8-aapl)
#EXTINF:16.666000,no-desc
Fragments(video=5999940,format=m3u8-aapl)
#EXTINF:14.300000,no-desc
Fragments(video=7499880,format=m3u8-aapl)
#EXTINF:16.667000,no-desc
Fragments(video=8786880,format=m3u8-aapl)
#EXTINF:16.667000,no-desc
Fragments(video=10286910,format=m3u8-aapl)
#EXTINF:16.666000,no-desc
Fragments(video=11786940,format=m3u8-aapl)
#EXTINF:16.667000,no-desc
Fragments(video=13286880,format=m3u8-aapl)
#EXTINF:16.667000,no-desc
Fragments(video=14786910,format=m3u8-aapl)
#EXTINF:16.666000,no-desc
Fragments(video=16286940,format=m3u8-aapl)
```


#### Manifest(audio_und,format=m3u8-aapl)
```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-ALLOW-CACHE:NO
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-TARGETDURATION:17
#EXT-X-PROGRAM-DATE-TIME:2020-02-12T09:52:26Z
#EXTINF:16.624989,no-desc
Fragments(audio_und=2911,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=736073,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=1471305,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=2206537,format=m3u8-aapl)
#EXTINF:16.625488,no-desc
Fragments(audio_und=2941769,format=m3u8-aapl)
#EXTINF:14.326712,no-desc
Fragments(audio_und=3674953,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=4306761,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=5041993,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=5777225,format=m3u8-aapl)
#EXTINF:16.625488,no-desc
Fragments(audio_und=6512457,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=7245641,format=m3u8-aapl)
#EXTINF:16.671926,no-desc
Fragments(audio_und=7980873,format=m3u8-aapl)
```

#### Manifest(textstream_und,format=m3u8-aapl)
```
#EXTM3U
#EXT-X-VERSION:4
#EXT-X-ALLOW-CACHE:NO
#EXT-X-MEDIA-SEQUENCE:0
#EXT-X-TARGETDURATION:4
#EXT-X-PROGRAM-DATE-TIME:2020-02-12T09:52:26Z
#EXTINF:2.944000,no-desc
Fragments(textstream_und=16896,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=64000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=128000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=192000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=256000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=320000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=384000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=448000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=512000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=576000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=640000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=704000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=768000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=832000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=896000,format=m3u8-aapl)
#EXTINF:2.000000,no-desc
Fragments(textstream_und=960000,format=m3u8-aapl)
#EXT-X-PROGRAM-DATE-TIME:2020-02-12T09:53:32.944Z
#EXT-X-DISCONTINUITY
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1088000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1152000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1216000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1280000,format=m3u8-aapl)
#EXT-X-PROGRAM-DATE-TIME:2020-02-12T09:53:52.944Z
#EXT-X-DISCONTINUITY
#EXTINF:2.000000,no-desc
Fragments(textstream_und=1408000,format=m3u8-aapl)
#EXT-X-PROGRAM-DATE-TIME:2020-02-12T09:53:58.944Z
#EXT-X-DISCONTINUITY
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1504000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1568000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1632000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1696000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1760000,format=m3u8-aapl)
#EXT-X-PROGRAM-DATE-TIME:2020-02-12T09:54:24.944Z
#EXT-X-DISCONTINUITY
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1920000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=1984000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2048000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2112000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2176000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2240000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2304000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2368000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2432000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2496000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2560000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2624000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2688000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2752000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2816000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2880000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=2944000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=3008000,format=m3u8-aapl)
#EXTINF:4.000000,no-desc
Fragments(textstream_und=3072000,format=m3u8-aapl)
```

Thanks,
Vishesh

**Comment by gesinger** [comment]
Thanks for the information @VisheshMittal . By the way, have you tested playback in Safari/iOS?

**Comment by VisheshMittal** [comment]
Hi @gesinger ,
I've tested the playback in Safari/iOS after overriding the native playback, and the behaviour is exactly similar to that in the latest stable release of video.js (i.e. there is no breaking change that I could notice with this PR).

Also, I noticed that there's the following console error that can be observed with the latest stable release of video.js and also visible with changes in this PR:
`TypeError: Argument 1 ('cue') to TextTrack.addCue must be an instance of TextTrackCue`

After some debugging, it turns out though the flow of execution is the same in most browsers, this error is specific to Safari, and occurs when `VTTCue` is added to the `TextTrack` (i.e. subtitleTextTrack).
From the specs: 
https://html.spec.whatwg.org/#text-track-cue
https://w3c.github.io/webvtt/#the-vttcue-interface

> Each text track cue has a corresponding TextTrackCue object (or more specifically, an object that inherits from TextTrackCue — for example, WebVTT cues use the VTTCue interface). A text track cue's in-memory representation can be dynamically changed through this TextTrackCue API. [WEBVTT]

Apart from the above, I couldn't notice any other things worth mentioning. I've run the testsuite many times, and it's working fine too. Please let me know if there's anything else that I should consider.
Thanks,
Vishesh

**Comment by gesinger** [comment]
Hey @VisheshMittal , thank you for the detailed response and debugging. I was primarily looking to see if your original stream plays in Safari's native player, without override native. Have you tested that case? And thanks again.

**Comment by VisheshMittal** [comment]
Ok. Yeah @gesinger , I've verified that the original stream plays in Safari's native player, without override native.

**Comment by VisheshMittal** [comment]
Hey @gesinger ,
Just wanted to check in if you see any other concerns. Please let me know if you'd need any other detail.
Thanks,
Vishesh

**Comment by gesinger** [comment]
Thanks @VisheshMittal .

At this time we're not able to merge this change. We appreciate the effort, and the documentation especially, and welcome further discussion on the matter, but right now we are working hard to make our syncing logic work as well as possible for all content which conforms to the specification. Because this content is out-of-spec, the fix is a workaround, which, while we may be able to consider it in the future, we are not prepared to maintain until we have cemented our synchronization across in-spec streams.

As for the native Safari player, I was mainly curious to see how it behaved, as, if it worked, then there is a solution working around out-of-spec content that may be considered in the future. And it may be the solution you provided. But at the moment we're not yet comfortable merging and maintaining it.

Thank you again.

**Pull request closed without merging**